src/main/tut/simple-algebraic-data-types.md

```diff
@@ -0,0 +1,375 @@
+def rho[A](x: A, u: Unit): A = x
+```
+...............
+```Haskell
+rho_inv :: a -> (a, ())
+rho_inv x = (x, ())
+```
+```tut:silent
+def rho_inv[A](x: A): (A, Unit) = (x, ())
+```
+...............
+```Haskell
+data Pair a b = P a b
+```
+```tut:silent
+sealed trait Pair[A, B]
+case class P[A, B](a: A, b: B) extends Pair[A, B]
+```
+...............
+```Haskell
+stmt :: Pair String Bool
+stmt = P "This statements is" False
+```
+```tut:silent
+val stmt: Pair[String, Boolean] =
+  P("This statements is", false)
+```
+...............
+```Haskell
+data Pair a b = Pair a b
+```
+```tut:silent
+case class Pair[A, B](a: A, b: B)
+```
+...............
+```Haskell
+stmt = (,) "This statement is" False
+```
+```tut:silent
+val stmt = ("This statement is", false)
+```
+...............
+```Haskell
+data Stmt = Stmt String Bool
+```
+```tut:silent
+case class Stmt(s: String, b: Boolean)
+```
+...............
+```Haskell
+startsWithSymbol :: (String, String, Int) -> Bool
+startsWithSymbol (name, symbol, _) = isPrefixOf symbol name
+```
+```tut:silent
+val startsWithSymbol: (String, String, Int) => Boolean =
+  (name, symbol, _) => name.startsWith(symbol)
+```
+...............
+```Haskell
+data Element = Element { name         :: String
+                       , symbol       :: String
+                       , atomicNumber :: Int }
+```
+```tut:silent
+case class Element(
+  name: String,
+  symbol: String,
+  atomicNumber: Int
+)
+```
+...............
+```Haskell
+tupleToElem :: (String, String, Int) -> Element
+tupleToElem (n, s, a) = Element { name = n
+                                , symbol = s
+                                , atomicNumber = a }
+```
+```tut:silent
+val tupleToElem: (String, String, Int) => Element =
+  (n, s, a) => Element(n, s, a)
+```
+...............
+```Haskell
+elemToTuple :: Element -> (String, String, Int)
+elemToTuple e = (name e, symbol e, atomicNumber e)
+```
+```tut:silent
+val elemToTuple: Element => (String, String, Int) =
+  e => (e.name, e.symbol, e.atomicNumber)
+```
+...............
+```Haskell
+atomicNumber :: Element -> Int
+```
+```scala
+val atomicNumber: Element => Int
+```
+...............
+```Haskell
+startsWithSymbol :: Element -> Bool
+startsWithSymbol e = isPrefixOf (symbol e) (name e)
+```
+```tut:silent
+val startsWithSymbol: Element => Boolean =
+  e => e.name startsWith e.symbol
+```
+...............
+```Haskell
+startsWithSymbol e = symbol e `isPrefixOf` name e
+```
+```tut:silent
+val startsWithSymbol: Element => Boolean =
+  e => e.name startsWith e.symbol
+```
+...............
+```Haskell
+data Either a b = Left a | Right b
+```
+```tut:silent
+sealed trait Either[+A, +B]
+case class Left[+A](a: A) extends Either[A, Nothing]
+case class Right[+B](b: B) extends Either[Nothing, B]
+```
+...............
+```Haskell
+data OneOfThree a b c = Sinistral a | Medial b | Dextral c
+```
+```tut:silent
+sealed trait OneOfThree[A, B, C]
+case class Sinistral[A](a: A) extends OneOfThree[A, Nothing, Nothing]
```

Comment:
Not sure if it's worth splitting these case classes into multiple lines e.g.
```Scala
case class Sinistral[A](a: A)
  extends OneOfThree[A, Nothing, Nothing]
```
since the length is fairly similar to the Haskell definition - what do you guys think?

Reply:
I am okay the way it is.